File: behavioral/typescript/interpreter_structured/boolean_expressions.ts
import { BooleanExpression, Context } from './interfaces';

export class VariableExpression implements BooleanExpression {
  private variableName: string;
  
  constructor(name: string) {
    this.variableName = name;
  }
  
  evaluate(context: Context): boolean {
    return context.getVariable(this.variableName) !== 0;
  }
}

export class ConstantExpression implements BooleanExpression {
  private value: boolean;
  
  constructor(value: boolean) {
    this.value = value;
  }
  
  evaluate(context: Context): boolean {
    return this.value;
  }
}

export class AndExpression implements BooleanExpression {
  private left: BooleanExpression;
  private right: BooleanExpression;
  
  constructor(left: BooleanExpression, right: BooleanExpression) {
    this.left = left;
    this.right = right;
  }
  
  evaluate(context: Context): boolean {
    return this.left.evaluate(context) && this.right.evaluate(context);
  }
}

export class OrExpression implements BooleanExpression {
  private left: BooleanExpression;
  private right: BooleanExpression;
  
  constructor(left: BooleanExpression, right: BooleanExpression) {
    this.left = left;
    this.right = right;
  }
  
  evaluate(context: Context): boolean {
    return this.left.evaluate(context) || this.right.evaluate(context);
  }
}

export class NotExpression implements BooleanExpression {
  private expression: BooleanExpression;
  
  constructor(expr: BooleanExpression) {
    this.expression = expr;
  }
  
  evaluate(context: Context): boolean {
    return !this.expression.evaluate(context);
  }
}

File: behavioral/dart/interpreter_structured/boolean_expressions.dart
import 'interfaces.dart';

class VariableExpression implements BooleanExpression {
  final String _variableName;
  
  VariableExpression(this._variableName);
  
  @override
  bool evaluate(Context context) {
    return context.getVariable(_variableName) != 0;
  }
}

class ConstantExpression implements BooleanExpression {
  final bool _value;
  
  ConstantExpression(this._value);
  
  @override
  bool evaluate(Context context) {
    return _value;
  }
}

class AndExpression implements BooleanExpression {
  final BooleanExpression _left;
  final BooleanExpression _right;
  
  AndExpression(this._left, this._right);
  
  @override
  bool evaluate(Context context) {
    return _left.evaluate(context) && _right.evaluate(context);
  }
}

class OrExpression implements BooleanExpression {
  final BooleanExpression _left;
  final BooleanExpression _right;
  
  OrExpression(this._left, this._right);
  
  @override
  bool evaluate(Context context) {
    return _left.evaluate(context) || _right.evaluate(context);
  }
}

class NotExpression implements BooleanExpression {
  final BooleanExpression _expression;
  
  NotExpression(this._expression);
  
  @override
  bool evaluate(Context context) {
    return !_expression.evaluate(context);
  }
}
